File: python/lexsieve/sink.py
"""Section 10.1: persistence contract. One immediate transaction checks
uniqueness, allocates seq, constructs/signs the receipt, inserts
decision/quarantine, and advances the head. No raw input, sanitized
content, model features, credential strings, or private signing keys are
persisted.
"""

from .errors import ClosedError
from .jcs import parse_json
from .schema import validate_decision, validate_signed_receipt

ZERO_HASH = "0" * 64


class MemorySink:
    """In-memory sink: same semantics as the SQLite sink, used by the
    conformance harness and embedders that provide their own storage."""

    def __init__(self):
        self.decisions = {}
        self.receipts = {}
        self.receipts_by_seq = {}
        self.heads = {}
        self.packs = {}
        self.active = {}
        self.quarantine = {}
        self.spool = {}
        # failure injection for fault-matrix tests
        self.fail_commit = False

    @staticmethod
    def _kg(t, g):
        return f"{t}/{g}"

    def open(self):
        pass

    def close(self):
        pass

    def get_decision(self, tenant, gateway, result_id):
        return self.decisions.get(f"{tenant}/{gateway}/{result_id}")

    def get_decision_by_binding(self, tenant, gateway, binding_hash):
        k = f"{tenant}/{gateway}"
        for d in self.decisions.values():
            if d["binding_hash"] == binding_hash and self.decisions.get(f"{k}/{d['result_id']}") is d:
                return d
        return None

    def get_receipt_by_id(self, receipt_id):
        return self.receipts.get(receipt_id)

    def get_chain_head(self, tenant, gateway):
        return self.heads.get(self._kg(tenant, gateway))

    def get_receipt_by_seq(self, tenant, gateway, seq):
        return self.receipts_by_seq.get(f"{tenant}/{gateway}/{seq}")

    def commit(self, tenant, gateway, result_id, input_hash, binding_hash, epoch, build):
        if self.fail_commit:
            raise ClosedError("STORAGE_UNAVAILABLE", "sink unavailable")
        dk = f"{tenant}/{gateway}/{result_id}"
        existing = self.decisions.get(dk)
        if existing:
            return {"status": "exists", "stored": existing}
        by_binding = self.get_decision_by_binding(tenant, gateway, binding_hash)
        if by_binding and by_binding["result_id"] != result_id:
            return {"status": "binding_conflict", "stored": by_binding}
        head = self.heads.get(self._kg(tenant, gateway), {"seq": 0, "hash": ZERO_HASH})
        seq = head["seq"] + 1
        built = build(seq, head["hash"])
        receipt, decision, quarantine = built["receipt"], built["decision"], built["quarantine"]
        self.receipts[receipt["body"]["receipt_id"]] = receipt
        self.receipts_by_seq[f"{tenant}/{gateway}/{seq}"] = receipt
        self.heads[self._kg(tenant, gateway)] = {"seq": seq, "hash": receipt["hash"]}
        self.decisions[dk] = {
            "result_id": result_id,
            "input_hash": input_hash,
            "binding_hash": binding_hash,
            "epoch": epoch,
            "receipt_id": receipt["body"]["receipt_id"],
            "decision": decision,
        }
        if quarantine:
            self.quarantine[quarantine["quarantine_id"]] = quarantine
        return {"status": "committed", "receipt": receipt}

    def put_pack(self, tenant, pack):
        self.packs[f"{tenant}/{pack['body']['pack_id']}/{pack['body']['serial']}"] = {
            "tenant_id": tenant,
            "pack_id": pack["body"]["pack_id"],
            "serial": pack["body"]["serial"],
            "hash": pack["hash"],
            "pack": pack,
        }

    def get_pack(self, tenant, pack_id, serial):
        return self.packs.get(f"{tenant}/{pack_id}/{serial}")

    def get_pack_by_hash(self, tenant, hash_):
        for p in self.packs.values():
            if p["tenant_id"] == tenant and p["hash"] == hash_:
                return p
        return None

    def get_max_pack_serial(self, tenant, pack_id):
        m = None
        for p in self.packs.values():
            if p["tenant_id"] == tenant and p["pack_id"] == pack_id:
                if m is None or p["serial"] > m:
                    m = p["serial"]
        return m

    def get_active_snapshot(self, tenant, gateway):
        return self.active.get(self._kg(tenant, gateway))

    def set_active_snapshot(self, tenant, gateway, rec):
        self.active[self._kg(tenant, gateway)] = rec

    def get_quarantine(self, ident):
        return self.quarantine.get(ident)

    def spool_put(self, row):
        self.spool[row["batch_id"]] = row

    def spool_due(self, now_ms, max_batches):
        due = [
            r
            for r in self.spool.values()
            if r["state"] == "pending" and r["next_attempt_ms"] <= now_ms
        ]
        due.sort(key=lambda r: r["expires_at_ms"])
        return due[:max_batches]

    def spool_update(self, row):
        self.spool[row["batch_id"]] = row

    def spool_count(self):
        return len(self.spool)

    def spool_bytes(self):
        return sum(len(r["json"]) for r in self.spool.values())

    def spool_evict_oldest(self, n):
        pending = [r for r in self.spool.values() if r["state"] == "pending"]
        pending.sort(key=lambda r: r["expires_at_ms"])
        evicted = 0
        for r in pending[:n]:
            del self.spool[r["batch_id"]]
            evicted += 1
        return evicted

    def receipt_range(self, tenant, gateway, from_seq, to_seq):
        out = []
        for s in range(from_seq, to_seq + 1):
            r = self.receipts_by_seq.get(f"{tenant}/{gateway}/{s}")
            if not r:
                break
            out.append(r)
        return out

    def receipt_count(self):
        return len(self.receipts)

    def verify_tail(self, tenant, gateway, n):
        head = self.heads.get(self._kg(tenant, gateway))
        if not head:
            return True
        last = self.receipts_by_seq.get(f"{tenant}/{gateway}/{head['seq']}")
        if not last or last["hash"] != head["hash"]:
            return False
        seq = head["seq"]
        expect_prev = last["body"]["prev_hash"]
        checked = 0
        while checked < n and seq > 1:
            prev = self.receipts_by_seq.get(f"{tenant}/{gateway}/{seq - 1}")
            if not prev or prev["hash"] != expect_prev:
                return False
            expect_prev = prev["body"]["prev_hash"]
            seq -= 1
            checked += 1
        if seq == 1 and expect_prev != ZERO_HASH:
            return False
        return True


# Re-validate stored rows defensively (they were validated at insert time).
def parse_stored_decision(json_str):
    return validate_decision(parse_json(json_str))


def parse_stored_receipt(json_str):
    return validate_signed_receipt(parse_json(json_str))

File: src/sink.ts
// Section 10.1: persistence contract. One immediate transaction checks
// uniqueness, allocates seq, constructs/signs the receipt, inserts
// decision/quarantine, and advances the head. No raw input, sanitized
// content, model features, credential strings, or private signing keys are
// persisted.

import { ClosedError } from './errors.ts';
import type { Decision, QuarantineMetadata, SignedReceipt } from './schema.ts';
import { parseJson, jcs, type JsonValue } from './jcs.ts';
import { validateDecision, validateSignedReceipt } from './schema.ts';

export interface StoredDecision {
  result_id: string;
  input_hash: string;
  binding_hash: string;
  epoch: number;
  receipt_id: string;
  decision: Decision;
}

export interface ChainHead {
  seq: number;
  hash: string;
}

export interface ActiveSnapshotRecord {
  epoch: number;
  snapshot: import('./schema.ts').Snapshot;
  pack_id: string;
  pack_serial: number;
  pack_expires_at_ms: number;
  pack_hash: string;
}

export interface StoredPack {
  tenant_id: string;
  pack_id: string;
  serial: number;
  hash: string;
  pack: import('./schema.ts').SignedPack;
}

export interface TelemetrySpoolRow {
  batch_id: string;
  json: string;
  state: 'pending' | 'sending' | 'acknowledged' | 'rejected' | 'expired';
  attempts: number;
  next_attempt_ms: number;
  expires_at_ms: number;
}

// build is invoked inside the commit transaction with the allocated chain
// position; it returns the signed receipt and optional quarantine metadata.
export type CommitBuild = (
  seq: number,
  prevHash: string,
) => { receipt: SignedReceipt; decision: Decision; quarantine: QuarantineMetadata | null };

export type CommitResult =
  | { status: 'committed'; receipt: SignedReceipt }
  | { status: 'exists'; stored: StoredDecision }
  | { status: 'binding_conflict'; stored: StoredDecision };

export interface Sink {
  open(): void;
  close(): void;
  getDecision(tenant: string, gateway: string, resultId: string): StoredDecision | null;
  getDecisionByBinding(tenant: string, gateway: string, bindingHash: string): StoredDecision | null;
  getReceiptById(receiptId: string): SignedReceipt | null;
  getChainHead(tenant: string, gateway: string): ChainHead | null;
  getReceiptBySeq(tenant: string, gateway: string, seq: number): SignedReceipt | null;
  commit(
    tenant: string,
    gateway: string,
    resultId: string,
    inputHash: string,
    bindingHash: string,
    epoch: number,
    build: CommitBuild,
  ): CommitResult;
  // pack registry
  putPack(tenant: string, pack: import('./schema.ts').SignedPack): void;
  getPack(tenant: string, packId: string, serial: number): StoredPack | null;
  getPackByHash(tenant: string, hash: string): StoredPack | null;
  getMaxPackSerial(tenant: string, packId: string): number | null;
  // active snapshot
  getActiveSnapshot(tenant: string, gateway: string): ActiveSnapshotRecord | null;
  setActiveSnapshot(tenant: string, gateway: string, rec: ActiveSnapshotRecord): void;
  // quarantine
  getQuarantine(id: string): QuarantineMetadata | null;
  // telemetry spool
  spoolPut(row: TelemetrySpoolRow): void;
  spoolDue(nowMs: number, maxBatches: number): TelemetrySpoolRow[];
  spoolUpdate(row: TelemetrySpoolRow): void;
  spoolCount(): number;
  spoolBytes(): number;
  spoolEvictOldest(n: number): number;
  // audit
  receiptRange(tenant: string, gateway: string, fromSeq: number, toSeq: number): SignedReceipt[];
  receiptCount(): number;
  verifyTail(tenant: string, gateway: string, n: number): boolean;
}

// ---------------------------------------------------------------------------
// In-memory sink: same semantics as the SQLite sink, used by the conformance
// harness and embedders that provide their own storage.
// ---------------------------------------------------------------------------

export class MemorySink implements Sink {
  private decisions = new Map<string, StoredDecision>();
  private receipts = new Map<string, SignedReceipt>();
  private receiptsBySeq = new Map<string, SignedReceipt>();
  private heads = new Map<string, ChainHead>();
  private packs = new Map<string, StoredPack>();
  private active = new Map<string, ActiveSnapshotRecord>();
  private quarantine = new Map<string, QuarantineMetadata>();
  private spool = new Map<string, TelemetrySpoolRow>();
  // failure injection for fault-matrix tests
  failCommit = false;

  private kg(t: string, g: string) {
    return `${t}/${g}`;
  }

  open(): void {}
  close(): void {}

  getDecision(tenant: string, gateway: string, resultId: string): StoredDecision | null {
    return this.decisions.get(`${tenant}/${gateway}/${resultId}`) ?? null;
  }

  getDecisionByBinding(tenant: string, gateway: string, bindingHash: string): StoredDecision | null {
    const k = `${tenant}/${gateway}`;
    for (const d of this.decisions.values()) {
      if (d.binding_hash === bindingHash && this.decisions.get(`${k}/${d.result_id}`) === d) return d;
    }
    return null;
  }

  getReceiptById(receiptId: string): SignedReceipt | null {
    return this.receipts.get(receiptId) ?? null;
  }

  getChainHead(tenant: string, gateway: string): ChainHead | null {
    return this.heads.get(this.kg(tenant, gateway)) ?? null;
  }

  getReceiptBySeq(tenant: string, gateway: string, seq: number): SignedReceipt | null {
    return this.receiptsBySeq.get(`${tenant}/${gateway}/${seq}`) ?? null;
  }

  commit(
    tenant: string,
    gateway: string,
    resultId: string,
    inputHash: string,
    bindingHash: string,
    epoch: number,
    build: CommitBuild,
  ): CommitResult {
    if (this.failCommit) throw new ClosedError('STORAGE_UNAVAILABLE', 'sink unavailable');
    const dk = `${tenant}/${gateway}/${resultId}`;
    const existing = this.decisions.get(dk);
    if (existing) return { status: 'exists', stored: existing };
    const byBinding = this.getDecisionByBinding(tenant, gateway, bindingHash);
    if (byBinding && byBinding.result_id !== resultId) {
      return { status: 'binding_conflict', stored: byBinding };
    }
    const head = this.heads.get(this.kg(tenant, gateway)) ?? { seq: 0, hash: '0'.repeat(64) };
    const seq = head.seq + 1;
    const { receipt, decision, quarantine } = build(seq, head.hash);
    this.receipts.set(receipt.body.receipt_id, receipt);
    this.receiptsBySeq.set(`${tenant}/${gateway}/${seq}`, receipt);
    this.heads.set(this.kg(tenant, gateway), { seq, hash: receipt.hash });
    this.decisions.set(dk, {
      result_id: resultId,
      input_hash: inputHash,
      binding_hash: bindingHash,
      epoch,
      receipt_id: receipt.body.receipt_id,
      decision,
    });
    if (quarantine) this.quarantine.set(quarantine.quarantine_id, quarantine);
    return { status: 'committed', receipt };
  }

  putPack(tenant: string, pack: import('./schema.ts').SignedPack): void {
    this.packs.set(`${tenant}/${pack.body.pack_id}/${pack.body.serial}`, {
      tenant_id: tenant,
      pack_id: pack.body.pack_id,
      serial: pack.body.serial,
      hash: pack.hash,
      pack,
    });
  }

  getPack(tenant: string, packId: string, serial: number): StoredPack | null {
    return this.packs.get(`${tenant}/${packId}/${serial}`) ?? null;
  }

  getPackByHash(tenant: string, hash: string): StoredPack | null {
    for (const p of this.packs.values()) {
      if (p.tenant_id === tenant && p.hash === hash) return p;
    }
    return null;
  }

  getMaxPackSerial(tenant: string, packId: string): number | null {
    let max: number | null = null;
    for (const p of this.packs.values()) {
      if (p.tenant_id === tenant && p.pack_id === packId) {
        if (max === null || p.serial > max) max = p.serial;
      }
    }
    return max;
  }

  getActiveSnapshot(tenant: string, gateway: string): ActiveSnapshotRecord | null {
    return this.active.get(this.kg(tenant, gateway)) ?? null;
  }

  setActiveSnapshot(tenant: string, gateway: string, rec: ActiveSnapshotRecord): void {
    this.active.set(this.kg(tenant, gateway), rec);
  }

  getQuarantine(id: string): QuarantineMetadata | null {
    return this.quarantine.get(id) ?? null;
  }

  spoolPut(row: TelemetrySpoolRow): void {
    this.spool.set(row.batch_id, row);
  }

  spoolDue(nowMs: number, maxBatches: number): TelemetrySpoolRow[] {
    return [...this.spool.values()]
      .filter((r) => r.state === 'pending' && r.next_attempt_ms <= nowMs)
      .sort((a, b) => a.expires_at_ms - b.expires_at_ms)
      .slice(0, maxBatches);
  }

  spoolUpdate(row: TelemetrySpoolRow): void {
    this.spool.set(row.batch_id, row);
  }

  spoolCount(): number {
    return this.spool.size;
  }

  spoolBytes(): number {
    let n = 0;
    for (const r of this.spool.values()) n += r.json.length;
    return n;
  }

  spoolEvictOldest(n: number): number {
    const pending = [...this.spool.values()]
      .filter((r) => r.state === 'pending')
      .sort((a, b) => a.expires_at_ms - b.expires_at_ms);
    let evicted = 0;
    for (const r of pending.slice(0, n)) {
      this.spool.delete(r.batch_id);
      evicted++;
    }
    return evicted;
  }

  receiptRange(tenant: string, gateway: string, fromSeq: number, toSeq: number): SignedReceipt[] {
    const out: SignedReceipt[] = [];
    for (let s = fromSeq; s <= toSeq; s++) {
      const r = this.receiptsBySeq.get(`${tenant}/${gateway}/${s}`);
      if (!r) break;
      out.push(r);
    }
    return out;
  }

  receiptCount(): number {
    return this.receipts.size;
  }

  verifyTail(tenant: string, gateway: string, n: number): boolean {
    const head = this.heads.get(this.kg(tenant, gateway));
    if (!head) return true;
    const last = this.receiptsBySeq.get(`${tenant}/${gateway}/${head.seq}`);
    if (!last || last.hash !== head.hash) return false;
    // check the last n links
    let seq = head.seq;
    let expectPrev = last.body.prev_hash;
    let checked = 0;
    while (checked < n && seq > 1) {
      const prev = this.receiptsBySeq.get(`${tenant}/${gateway}/${seq - 1}`);
      if (!prev || prev.hash !== expectPrev) return false;
      expectPrev = prev.body.prev_hash;
      seq--;
      checked++;
    }
    if (seq === 1 && expectPrev !== '0'.repeat(64)) return false;
    return true;
  }
}

// Re-validate stored rows defensively (they were validated at insert time).
export function parseStoredDecision(json: string): Decision {
  return validateDecision(parseJson(json));
}

export function parseStoredReceipt(json: string): SignedReceipt {
  return validateSignedReceipt(parseJson(json));
}

export { jcs };
export type { JsonValue };
